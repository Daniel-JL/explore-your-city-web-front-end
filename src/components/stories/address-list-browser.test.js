import React from 'react';
import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import userEvent from '@testing-library/user-event'
import {
  Router,
} from 'react-router-dom';
import { createMemoryHistory } from 'history';
import '@testing-library/jest-dom/extend-expect';
import AddressListBrowser from './address-list-browser';

const mockHandleLeftRightButtonClick = jest.fn();

describe('Address list browser', () => {
  it('should render correctly', () => {
    const wrapper = renderer.create(<AddressListBrowser addressList={[]}/>).toJSON();

    expect(wrapper).toMatchSnapshot();
  });

  it('should display items as they are added to address list', () => {
    const history = createMemoryHistory();
    const { getByRole, rerender } = render(
      <Router history={history}>
        <AddressListBrowser
          addressList={[
            {
              address: 'first-item',
              locationStory: 'first-story',
            },
          ]}
        />
      </Router>,
    );

    const addressBox = screen.getByRole('textbox', { name: /address-field/i });
    const storyBox = screen.getByRole('textbox', { name: /location-story/i });

    expect(addressBox.value).toBe('first-item');
    expect(storyBox.value).toBe('first-story');

    rerender(
      <AddressListBrowser
        addressList={[
          {
            address: 'first-item',
            locationStory: 'first-story',
          },
          {
            address: 'second-item',
            locationStory: 'second-story',
          },
        ]}
      />,
    );

    const addressBoxRerender = screen.getByRole('textbox', { name: /address-field/i });
    const storyBoxRerender = screen.getByRole('textbox', { name: /location-story/i });

    expect(addressBoxRerender.value).toBe('second-item');
    expect(storyBoxRerender.value).toBe('second-story');
  });

  it('should not change the display data if the max or min of the array is reached', () => {
    const history = createMemoryHistory();
    const { getByRole, rerender } = render(
      <Router history={history}>
        <AddressListBrowser
          addressList={[
            {
              address: '',
              locationStory: '',
            },
            {
              address: 'first-item',
              locationStory: 'first-story',
            },
            {
              address: 'second-item',
              locationStory: 'second-story',
            },
            {
              address: 'third-item',
              locationStory: 'third-story',
            },
          ]}
          handleLeftRightButtonClick={mockHandleLeftRightButtonClick}
        />
      </Router>,
    );

    const addressBox = screen.getByRole('textbox', { name: /address-field/i });
    const storyBox = screen.getByRole('textbox', { name: /location-story/i });
    const leftArrowButton = screen.getByRole('button', { name: /left-arrow/i });
    const rightArrowButton = screen.getByRole('button', { name: /right-arrow/i });

    expect(addressBox.value).toBe('third-item');
    expect(storyBox.value).toBe('third-story');

    userEvent.click(rightArrowButton);

    expect(addressBox.value).toBe('third-item');
    expect(storyBox.value).toBe('third-story');

    userEvent.click(leftArrowButton);

    expect(addressBox.value).toBe('second-item');
    expect(storyBox.value).toBe('second-story');

    userEvent.click(leftArrowButton);

    expect(addressBox.value).toBe('first-item');
    expect(storyBox.value).toBe('first-story');

    userEvent.click(leftArrowButton);

    expect(addressBox.value).toBe('first-item');
    expect(storyBox.value).toBe('first-story');

    userEvent.click(rightArrowButton);

    expect(addressBox.value).toBe('second-item');
    expect(storyBox.value).toBe('second-story');
  });

  // it('should remove an item from the address list when the delete button is pressed', () => {
  //   //  TODO
  //   //  This should be a part of integration tests as the address list value
  //   //  changes based on interactions with the contribution page
  //   const history = createMemoryHistory();
  //   const { getByRole, rerender } = render(
  //     <Router history={history}>
  //       <AddressListBrowser
  //         addressList={[
  //           {
  //             address: '',
  //             locationStory: '',
  //           },
  //           {
  //             address: 'first-item',
  //             locationStory: 'first-story',
  //           },
  //           {
  //             address: 'second-item',
  //             locationStory: 'second-story',
  //           },
  //         ]}
  //       />
  //     </Router>,
  //   );

  //   const addressBox = screen.getByRole('textbox', { name: /address-field/i });
  //   const storyBox = screen.getByRole('textbox', { name: /location-story/i });
  //   const trashButton = screen.getByRole('button', { name: /trash-button/i });

  //   expect(addressBox.value).toBe('second-item');
  //   expect(storyBox.value).toBe('second-story');

  //   // userEvent.click(trashButton);

  //   // console.log('before rerender');
  //   // rerender(
  //   //   <AddressListBrowser
  //   //     addressList={[
  //   //       {
  //   //         address: '',
  //   //         locationStory: '',
  //   //       },
  //   //       {
  //   //         address: 'first-item',
  //   //         locationStory: 'first-story',
  //   //       },
  //   //     ]}
  //   //   />,
  //   // );

  //   // expect(addressBox.value).toBe('first-item');
  //   // expect(storyBox.value).toBe('first-story');




  // });

  it('should hide the story description box when the hide/show button is pressed', () => {
    const history = createMemoryHistory();
    const { getByRole } = render(
      <Router history={history}>
        <AddressListBrowser
          addressList={[
            {
              address: '',
              locationStory: '',
            },
            {
              address: 'first-item',
              locationStory: 'first-story',
            },
          ]}
        />
      </Router>,
    );

    const addressBox = screen.getByRole('textbox', { name: /address-field/i });
    const storyBox = screen.getByRole('textbox', { name: /location-story/i });
    const hideTextButton = screen.getByRole('button', { name: /hide-text-button/i });

    expect(addressBox.value).toBe('first-item');
    expect(storyBox.value).toBe('first-story');
    expect(storyBox).toHaveStyle('display: block');

    userEvent.click(hideTextButton);

    expect(storyBox).toHaveStyle('display: none');
  });
});
