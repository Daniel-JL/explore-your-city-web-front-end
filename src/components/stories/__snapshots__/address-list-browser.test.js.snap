// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Address list browser should render correctly 1`] = `
<div
  className="sc-fubCfw jkwkKq"
>
  <div
    className="sc-pFZIQ fZtmim"
  >
    <button
      aria-label="left-arrow"
      className="sc-bdfBwQ dDwjPX"
      onClick={[Function]}
    >
      Left
    </button>
    <button
      aria-label="right-arrow"
      className="sc-bdfBwQ dDwjPX"
      onClick={[Function]}
    >
      Right
    </button>
    <button
      aria-label="trash-button"
      className="sc-bdfBwQ sc-dlfnbm dDwjPX jGyOBP"
      id="trash-button"
      onClick={[Function]}
    >
      Trash
    </button>
    <button
      aria-label="hide-text-button"
      className="sc-bdfBwQ dDwjPX"
      id="hide-text-button"
      onClick={[Function]}
    >
      Hide/Show Text
    </button>
  </div>
  <input
    aria-label="address-field"
    className="sc-jrAGrp eUrxvH"
    id="address-field"
    name="address-field"
    readOnly={true}
    type="text"
    value=""
  />
  <textarea
    aria-label="location-story"
    className="sc-kEjbxe clHPVE"
    id="location-story"
    readOnly={true}
    style={
      Object {
        "display": "block",
      }
    }
    value=""
  />
</div>
`;
